boards: stm32: make flashing with stm32cubeprogrammer available & default on all boards

ajarmouni-st opened the pull request:
Add stm32cubeprogrammer where it is not available, & make it the default flashing method. Update doc to list flashing options.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/75237

Edit: to give more context, right now the west runner that's selected when calling `west flash` without `--runner ...`option is kind of arbitrary & depends on which runner is first in `board.cmake`.

erwango commented:
> @kartben Do you think this can go in v3.7.0?

I prefer avoiding such bulk change in RC2 context. Doc update is welcome though.

glenn-andrews commented:
Does this now require people to make an ST account to download STM32CubeProg?

OpenOCD and JLink doesn't need an account on a corporation to download.

glenn-andrews commented:
> @glenn-andrews it can be downloaded without an account in guest mode, but an email address is needed to receive the link.
> 
> In any case, this PR aims to unify & simplify the user experience when the `--runner` option is not specified. We can put OpenOCD as default, but then Zephyr SDK does not provide a pre-built OpenOCD binary on Windows, & it takes months anyway to add support for the latest STM32 SoCs in the main upstream OpenOCD branch. 

I'd prefer streamlining the installation instructions for OpenOCD. You can't debug with STM32CubeProg, so you need to install something else anyway.

What do we do for debug support until new STM32 SoC are added?

JarmouniA commented:
> What do we do for debug support until new STM32 SoC are added?

You can debug with ST-link Server (https://www.st.com/en/development-tools/st-link-server.html#overview) or ST-OpenOCD, both are included with STM32CubeCLT (https://www.st.com/en/development-tools/stm32cubeclt.html).

glenn-andrews commented:
> You can debug with ST-link Server (https://www.st.com/en/development-tools/st-link-server.html#overview) or ST-OpenOCD, both are included with STM32CubeCLT (https://www.st.com/en/development-tools/stm32cubeclt.html).

So should we be documenting to download that package and how to set it up, rather than just the programmer?

JarmouniA committed:
doc: releases: migration-guide-4.0: note about stm32 boards flashing

STM32CubeProgrammer is now selected as the default runner for all STM32
official boards when 'west flash' is invoked.

Signed-off-by: Abderrahmane Jarmouni <git@jarmouni.me>

JarmouniA committed:
doc: flash_debug: document stm32cubeprogrammer

Document STM32CubeProgrammer installation & usage with Zephyr.

Signed-off-by: Abderrahmane Jarmouni <git@jarmouni.me>

glenn-andrews commented:
> > So should we be documenting to download that package and how to set it up, rather than just the programmer?
> 
> Right now, `west debug` supports only OpenOCD, but I think support should be added for ST-link GDB server as well, esp. since, I believe, there are STM32 targets that can only be debugged with the later, like CM4 core on the dual-core STM32H747I.

How heavy a lift is fixing `west debug`? Right now to use STM32 you install OpenOCD, unless you're working with newer MCUs. After this you'll need to install both OpenOCD and STM32CubeProg to flash and debug (or always specify the OpenOCD runner for flashing).

A one-stop-shop of installing STM32CubeCLT would be preferable than having to install multiple tool packages.

erwango commented:
II approved but I forgot the most important: Nice to see you around @JarmouniA :)

glenn-andrews commented on doc/releases/migration-guide-4.0.rst:
```diff
@@ -27,6 +27,13 @@ Boards
+=====
+
+* On all official STM32 boards, `west flash` selects STM32CubeProgrammer as the default west runner.
+  If you want to enforce the selection of another runner like OpenOCD or pyOCD for flashing, you should
```
Is there a way in west to permanently select a preferred runner?

i.e. if you _really_ want to not install STM32CubeProg, can you configure west such that it always selects OpenOCD for flashing when no runner is given on the command line?

erwango replied:
It is possible using twister and map file.
Using `west`, I thought it was possible using [configuration options](https://docs.zephyrproject.org/latest/develop/west/build-flash-debug.html#configuration-options), which is unfortunately not the case apparently, but I think can be added.

str4t0m commented on boards/st/nucleo_g474re/doc/index.rst:
```diff
@@ -207,6 +207,17 @@ Flashing
+.. code-block:: console
+
+   $ west flash --runner openocd
+   $ west flash --runner jlink
```
Additionally to stm32cubeprogrammer, it only has pyocd and openocd.
But jlink could be added in a follow up PR.

str4t0m commented on boards/st/stm32f103_mini/doc/index.rst:
```diff
@@ -162,3 +162,6 @@ References
 
 .. _STM32F103 data sheet:
    https://www.st.com/resource/en/datasheet/stm32f103rc.pdf
+
+.. _STM32CubeProgrammer:
+   https://www.st.com/en/development-tools/stm32cubeprog.html
```
NIT: I think that link is never referenced.

str4t0m commented on boards/st/nucleo_g431rb/doc/index.rst:
```diff
@@ -193,25 +193,16 @@ Flashing
-
-   $ west flash --openocd /usr/local/bin/openocd
+   $ west flash --runner openocd
+   $ west flash --runner jlink
```
NIT: Jlink hasn't been added for this board yer.

str4t0m commented:
Will need a rebase to resolve a conflict in the  migration guide.

JarmouniA commented:
Rebase

JarmouniA committed:
boards: move stm32f103_mini to others

move stm32f103_mini board from ST to Others as it is not an official
ST board.

Signed-off-by: Abderrahmane Jarmouni <git@jarmouni.me>

JarmouniA commented:
@str4t0m your remarks have been addressed.

JarmouniA committed:
boards: stm32: add flashing with stm32cubeprog for all

Add STM32CubeProgrammer where it is not available, & make it
default flashing method by placing it 1st in board.cmake

Before this change, the west runner that's selected when calling
'westflash' without the '--runner ...' option is kind of arbitrary
& depends on which runner is 1st in board.cmake file.

Another reason this change is needed is that STM32CubeProg is the only
available way to check/modify Option Bytes, and to flash code to
external Flash on many STM32 boards.

This change does not affect 'west debug' behavior, which defaults to
the debugger (OpenOCD/JLink) that is included first in board.cmake

Signed-off-by: Abderrahmane Jarmouni <git@jarmouni.me>

JarmouniA committed:
boards: stm32: doc: update flashing section

Update all STM32 official boards Doc Flashing section.
Correct some typos and remove outdated information.

Signed-off-by: Abderrahmane Jarmouni <git@jarmouni.me>
